clientconn.go

```diff
@@ -133,12 +140,35 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 		if closeTransport {
 			t.Close()
 		}
-		newTransport, err := transport.NewClientTransport(cc.dopts.protocol, cc.target, cc.dopts.authOptions)
+		// Adjust timeout for the current try.
+		dopts := cc.dopts
+		if dopts.Timeout < 0 {
+			cc.Close()
+			return ErrClientConnTimeout
+		}
+		if dopts.Timeout > 0 {
+			dopts.Timeout -= time.Since(start)
+			if dopts.Timeout <= 0 {
+				cc.Close()
+				return ErrClientConnTimeout
+			}
+		}
+		newTransport, err := transport.NewClientTransport(cc.target, &dopts)
 		if err != nil {
-			// TODO(zhaoq): Record the error with glog.V.
+			if netErr, ok := err.(net.Error); ok && netErr.Timeout() {
```

Comment:
More particularly, a net.Error being a timeout does not guarantee that the full cc.dopts.Timeout has elapsed. It might be a sub-second failure due to name resolution, for instance.